packages/driver/src/cy/commands/actions/attachFile.ts

```diff
@@ -6,23 +6,111 @@ import $errUtils from '../../../cypress/error_utils'
 import $actionability from '../../actionability'
 import { addEventCoords, dispatch } from './trigger'
 
-const createFileList = (files: Cypress.FileReferenceObject[]): FileList => {
+/* dropzone.js relies on an experimental, nonstandard API, webkitGetAsEntry().
+ * https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry
+ * The behavior here varies between browsers, and though unpleasant, we have to attempt
+ * to replicate it as far as possible.
+ *
+ * In Firefox, attempting to set webkitGetAsEntry() fails - but this is fine, because
+ * Firefox also returns a useful value from item.webkitGetAsEntry().
+ *
+ * In webkit browsers however, item.webkitGetAsEntry() returns null, because our File()
+ * instances don't actually correspond with a file on disk. But the property is writeable,
+ * so we're able to override it with our own implementation that returns a proper object.
+ */
+const tryMockWebkit = (item) => {
+  try {
+    item.webkitGetAsEntry = () => {
+      return {
+        isFile: true,
+        file: (callback) => callback(item.getAsFile()),
+      }
+    }
+  } catch (e) {
+    // We're in Firefox, this is fine.
+  }
+
+  return item
+}
+
+const createDataTransfer = (files: Cypress.FileReferenceObject[]): DataTransfer => {
   const dataTransfer = new DataTransfer()
 
   files.forEach(({ contents, fileName, lastModified, mimeType }) => {
-    const file = new File([contents], fileName || '', { lastModified, type: mimeType })
+    dataTransfer.items.add(new File([contents], fileName || '', {
+      lastModified,
+      type: mimeType,
+    }))
+  })
+
+  const oldItems = dataTransfer.items
 
-    dataTransfer.items.add(file)
+  // dataTransfer.items is a getter, and it - and the items read from its
+  // returned DataTransferItemList - cannot be assigned to. DataTransferItemLists
+  // also cannot be constructed, so we have to use an array instead.
+  Object.defineProperty(dataTransfer, 'items', {
+    get () {
+      return _.map(oldItems, tryMockWebkit)
+    },
   })
 
-  return dataTransfer.files
+  // When a real user drags file(s) over an element, dataTransfer.types always contains `['Files']`.
+  // In Firefox however, the 'types' array is not properly populated when we construct our dataTransfer.
+  if (files.length !== 0 && dataTransfer.types.length === 0) {
```

Comment:
Agains should we be verifying this is Firefox?

Reply:
I've always been a fan of duck-typing - it looks like a duck, it quacks like a duck, it's a duck. Or to use more formal language, feature detection rather than browser detection.

In this case - if it should have a types array but it's empty, we add one. If some new mobile browser comes out wrapping Firefox, or implementing its own behavior similar to firefox, a feature check will keep working, but a browser-check might fail.